text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,673 @@
+
+# Current State of Art
+## Web Components community
+There is no well established standard or format for declaratively defining a web component. [custom-elements-json](https://github.com/webcomponents/custom-elements-json) is making an attempt to define a standard. It is in a very early stage. However there is a ecosystem developing around this standard. 
+[web-component-analyzer](https://www.npmjs.com/package/web-component-analyzer) can analyze web components written in vanilla javascript and other popular web component libraries. It also supports custom-elements-json as an output format.
+
+## LWC Platform
+Currently, LWC platform gathers metadata as part of the compilation process. This metadata shape and implementation is currently private. This RFC is an attempt to make the metadata shape public as a first step. This [document](https://salesforce.quip.com/DdncANrJA0ko)<sup>*</sup> captures the information collected and the metadata shape currently gathered. Here is a summary of problems with the current implementation:
+* Replicated type system in java and javascript that is maintained manually
+* Dependency(reference) analysis is done outside of metadata gathering
+* Change management to accomodate new metadata is cumbersome because the implementation is spread across two repos
+
+# Detailed Design
+
+## Format
+Metadata shape of the various objects we wish to analyse will be represented using in [JSON Schema](https://json-schema.org/) format. This is also augmented with a TypeScript definition of the shape.
+
+[JSON](https://tools.ietf.org/html/rfc7159) is the IETF approved standard for representing objects. [JSON Schema](https://json-schema.org/) is on the path to become an IETF approved standard for describing JSON.
+
+This format was also influenced by standardization choices made with in Salesforce UI platform.
+
+## Entities that will be analyzed
+These are the entities that will be analzed and metadata gathered about:
+* HTML Template file
+    * Custom element references
+        * Tag name
+        * Attributes
+        * Slot content - default and named slots
+    * Static resource
+        * Image urls
+        * Javascript resource
+        * CSS resource
+    * Dynamic components
+    * Template directives
+        * for:each
+        * if
+        * iterator
+        * key
+        * lwc:dom="manual"
+    * Event handlers
+    * Slots
+* Javascript file
+    * Component classes
+        * Inheritance
+        * Public & private properties
+        * Decorators
+        * getters/setters
+        * Public methods
+    * Imports
+        * Builtin modules
+        * External modules
+        * Salesforce scoped modules
+        * Other modules(components and libraries)
+        * Relative imports
```

Comment:
what is the use case for relative imports? any bundling tools, such as rollup, will inline relative imports during build time.

Reply:
Given a module, the metadata gathering collects information about every import statement in the module. It is agnostic of what build tools are used post metadata collection. That is not the concern of the metadata collector. If the module has a relative import, that information should be surfaced.

Reply:
Why do we want to surface internal information like that. No one outside the module should be able to see or access that information. I would argue that we should not surface that.

Reply:
we will need information about salesforce scoped imports in order to pre-cache resources (contentAssetUrl, resourceUrl, Java resources)

Reply:
Once I put forward my RFC/proposal for per-file pipeline this will make more sense.